ops/storage.py

```diff
@@ -0,0 +1,263 @@
+        self.store_notice_list(notice_list)
+
+    def drop_notice(self, event_path, observer_path, method_name):
+        notice_list = self.load_notice_list()
+        notice_list.remove([event_path, observer_path, method_name])
+        self.store_notice_list(notice_list)
+
+    def notices(self, event_path):
+        notice_list = self.load_notice_list()
+        if event_path:
+            notice_list = list(filter(lambda row: row[0] == event_path, notice_list))
+        for row in notice_list:
+            yield tuple(row)
+
+    def load_notice_list(self):
+        serialized_notices = self._backend.get("#notices#")
+        if serialized_notices is None:
+            return []
+        return serialized_notices
+
+    def store_notice_list(self, notice_list):
+        self.self._backend.set("#notices#", notice_list)
+
+
+class _SimpleLoader(yaml.SafeLoader):
+    """Handle a couple basic python types.
+
+    yaml.SafeLoader can handle all the basic int/float/dict/set/etc that we want. The only one
+    that it *doesn't* handle is tuples. We don't want to support arbitrary types, so we just
+    subclass SafeLoader and add tuples back in.
+    """
+    # Taken from the example at:
+    # https://stackoverflow.com/questions/9169025/how-can-i-add-a-python-tuple-to-a-yaml-file-using-pyyaml
+
+
+_SimpleLoader.construct_python_tuple = yaml.Loader.construct_python_tuple
+_SimpleLoader.add_constructor(
+    u'tag:yaml.org,2002:python/tuple',
+    _SimpleLoader.construct_python_tuple)
+
+
+class _SimpleDumper(yaml.SafeDumper):
+    """Add types supported by 'marshal'
+
+    YAML can support arbitrary types, but that is generally considered unsafe (like pickle). So
+    we want to only support dumping out types that are safe to load.
+    """
+
+
+_SimpleDumper.represent_tuple = yaml.Dumper.represent_tuple
+_SimpleDumper.add_representer(tuple, _SimpleDumper.represent_tuple)
+
+
+class _JujuStorageBackend:
+    """Implements the interface from the Operator framework to Juju's state-get/set/etc."""
+
+    @staticmethod
+    def is_available() -> bool:
+        """Check if Juju state storage is available.
+
+        This checks if there is a 'state-get' executable in PATH.
+        """
+        p = shutil.which('state-get')
+        return p is not None
+
+    def set(self, key: str, value: typing.Any) -> None:
+        """Set a key to a given bytes value.
+
+        Args:
+            key: The string key that will be used to find the value later
+            value: Arbitrary bytes that will be returned by get().
+        Raises:
+            CalledProcessError: if 'state-set' returns an error code.
+        """
+        encoded_value = yaml.dump(value, Dumper=_SimpleDumper)
+        content = yaml.safe_dump({key: encoded_value}, encoding='utf-8',
+                                 default_style='|', default_flow_style=False)
+        # Note: 'capture_output' would be good here, but was added in Python 3.7
```

Comment:
`capture_output` is just shorthand for `stdout=PIPE, stderr=PIPE`:

https://github.com/python/cpython/blob/c6b292cdeee689f0bfac6c1e2c2d4e4e01fa8d9e/Lib/subprocess.py#L494-L499